hphp/runtime/ext/std/ext_std_file.cpp

```diff
@@ -588,9 +588,21 @@ Variant HHVM_FUNCTION(file_put_contents,
                       int flags /* = 0 */,
                       const Variant& context /* = null */) {
   CHECK_PATH(filename, 1);
+
+  char mode[3] = "wb";
+  if (flags & PHP_FILE_APPEND) {
+    mode[0] = 'a';
+  } else if (flags & LOCK_EX) {
+    // Open in "create" mode (writing only, create if needed, no truncate)
+    // so that the file is not modified before we attempt to aquire the
+    // requested lock.
+    mode[0] = 'c';
```

Comment:
Yeah, it's special PHP magic. It looks to me like [PlainFile::open](https://github.com/facebook/hhvm/blob/master/hphp/runtime/base/plain-file.cpp#L85-L95) at least does the right thing. I'm not 100% sure about all other stream types.

This patch did fix the reproduction case given in https://bugs.php.net/bug.php?id=43182 for me in local testing but I'm not sure how to translate a test that needs multiple processes into the HHVM test system.